internal/gengapic/paging.go

```diff
@@ -86,82 +87,75 @@ func (g *generator) iterTypeOf(elemField *descriptor.FieldDescriptorProto) (*ite
+		return nil, nil, errors.E(nil, "cannot find message type %q, malformed descriptor?", m.GetInputType())
 	}
 	inMsg, ok := inType.(*descriptor.DescriptorProto)
 	if !ok {
-		return nil, errors.E(nil, "expected %q to be message type, found %T", m.GetInputType(), inType)
+		return nil, nil, errors.E(nil, "cannot find message type %q, malformed descriptor?", m.GetInputType())
 	}
 
 	outType := g.descInfo.Type[m.GetOutputType()]
 	if outType == nil {
-		return nil, errors.E(nil, "cannot find message type %q, malformed descriptor?", m.GetOutputType())
+		return nil, nil, errors.E(nil, "cannot find message type %q, malformed descriptor?", m.GetOutputType())
 	}
 	outMsg, ok := outType.(*descriptor.DescriptorProto)
 	if !ok {
-		return nil, errors.E(nil, "expected %q to be message type, found %T", m.GetOutputType(), outType)
+		return nil, nil, errors.E(nil, "cannot find message type %q, malformed descriptor?", m.GetOutputType())
 	}
 
-	for _, f := range inMsg.Field {
-		if f.GetName() == "page_size" && f.GetType() == descriptor.FieldDescriptorProto_TYPE_INT32 {
-			hasSize = true
-		}
-		if f.GetName() == "page_token" && f.GetType() == descriptor.FieldDescriptorProto_TYPE_STRING {
-			hasToken = true
-		}
-	}
-	for _, f := range outMsg.Field {
-		if f.GetName() == "next_page_token" && f.GetType() == descriptor.FieldDescriptorProto_TYPE_STRING {
-			hasNextToken = true
-		}
-		if f.GetLabel() == descriptor.FieldDescriptorProto_LABEL_REPEATED {
-			elemFields = append(elemFields, f)
+	hasPageToken := false
+	for _, f := range inMsg.GetField() {
+		if (f.GetName() == "page_size" || f.GetName() == "max_results") && f.GetType() == descriptor.FieldDescriptorProto_TYPE_INT32 {
+			if pageSizeField != nil {
+				return nil, nil, errors.E(nil, "found both page_size and max_results fields in message %q", m.GetInputType())
+			}
+			pageSizeField = f
+			continue
 		}
+
+		hasPageToken = hasPageToken || (f.GetName() == "page_token" && f.GetType() == descriptor.FieldDescriptorProto_TYPE_STRING)
 	}
 
-	// If any of the required pagination fields are missing or if the response
-	// has no repeated fields, it is not a fully conforming paginated method.
-	if !hasSize || !hasToken || !hasNextToken || len(elemFields) == 0 {
-		return nil, nil
+	if !hasPageToken || pageSizeField == nil {
+		// Not an error, just not paginated
+		return nil, nil, nil
 	}
 
-	if len(elemFields) > 1 {
-		// ensure that the first repeated field visited has the lowest field number
-		// according to https://aip.dev/4233, and use that one.
-		min := elemFields[0].GetNumber()
-		for _, elem := range elemFields {
-			if elem.GetNumber() < min {
-				return nil, fmt.Errorf("%s: can't pick repeated field in %s based on aip.dev/4233", *m.Name, outType.GetName())
+	hasNextPageToken := false
+	for _, f := range outMsg.GetField() {
+		if f.GetLabel() == descriptor.FieldDescriptorProto_LABEL_REPEATED {
```

Comment:
You're not checking for map responses yet? I would either
(a) add a TODO and change the PR description to say you're deferring map-method detection or
(b) detect map responses here and now and add an error/TODO in the paging iterator for map responses

Reply:
Before checking the options struct for diregapic mode, there's not a lot to do for detecting map responses here. They are also repeated; the main difference would be that we need to generate a custom pair type and use that for the iterator.